storage/src/main/java/org/apache/kafka/storage/internals/log/LogCleanerManager.java

```
@@ -0,0 +1,800 @@
+                .flatMap(Optional::stream)
+                .min(Long::compare)
+                .orElseThrow(() -> new IllegalStateException("No uncleanable offset found"));
+
+        LOG.debug("Finding range of cleanable offsets for log={}. Last clean offset={} " +
+                        "now={} => firstDirtyOffset={} firstUncleanableOffset={} activeSegment.baseOffset={}",
+                log.name(), lastCleanOffset, now, firstDirtyOffset, firstUncleanableDirtyOffset, log.activeSegment().baseOffset());
+
+        return new OffsetsToClean(firstDirtyOffset, Math.max(firstDirtyOffset, firstUncleanableDirtyOffset), forceUpdateCheckpoint);
+    }
+
+    /**
+     * Given the first dirty offset and an uncleanable offset, calculates the total cleanable bytes for this log.
+     *
+     * @return the biggest uncleanable offset and the total amount of cleanable bytes
+     */
+    public static Map.Entry<Long, Long> calculateCleanableBytes(UnifiedLog log, long firstDirtyOffset, long uncleanableOffset) {
+        List<LogSegment> nonActiveSegments = log.nonActiveLogSegmentsFrom(uncleanableOffset);
+        LogSegment firstUncleanableSegment = nonActiveSegments.isEmpty() ? log.activeSegment() : nonActiveSegments.get(0);
+        long firstUncleanableOffset = firstUncleanableSegment.baseOffset();
+        long cleanableBytes = log.logSegments(Math.min(firstDirtyOffset, firstUncleanableOffset), firstUncleanableOffset).stream()
+                .mapToLong(LogSegment::size)
+                .sum();
+
+        return Map.entry(firstUncleanableOffset, cleanableBytes);
+    }
+
+    private static Optional<Long> findFirstUncleanableSegment(UnifiedLog log, long firstDirtyOffset, long now, long minCompactionLagMs) {
+        List<LogSegment> dirtyNonActiveSegments = log.nonActiveLogSegmentsFrom(firstDirtyOffset);
+        return dirtyNonActiveSegments.stream()
+                .filter(segment -> {
+                    boolean isUncleanable;
+                    try {
```

```java
                    try {
                        boolean isUncleanable = segment.largestTimestamp() > now - minCompactionLagMs;
                        LOG.debug("Checking if log segment may be cleaned: log='{}' segment.baseOffset={} " +
                                        "segment.largestTimestamp={}; now - compactionLag={}; is uncleanable={}",
                                log.name(), segment.baseOffset(), segment.largestTimestamp(), now - minCompactionLagMs, isUncleanable);
                        return isUncleanable;
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
```

---

For another, should we throw `LogCleaningException` instead of `RuntimeException`? `LogCleaner#tryCleanFilthiestLog` does not catch all exception, so throwing non-`LogCleaningException` can cease the thread of `LogCleaner`. That could be a terrible issue as users have to reconfigure the `LogCleaner` to bring them back. Maybe making `LogCleaner#tryCleanFilthiestLog` handle all exceptions is more acceptable.

```scala
    private def tryCleanFilthiestLog(): Boolean = {
      try {
        cleanFilthiestLog()
      } catch {
        case e: LogCleaningException =>
          warn(s"Unexpected exception thrown when cleaning log ${e.log}. Marking its partition (${e.log.topicPartition}) as uncleanable", e)
          cleanerManager.markPartitionUncleanable(e.log.parentDir, e.log.topicPartition)

          false
      }
    }
```

---

The `findFirstUncleanableSegment` method is used in the `cleanableOffsets` method, which in turn is called in the `grabFilthiestCompactedLog` method in a try - catch block where all errors are wrapped in a LogCleaningException, so it seems everything will work as before.
Correct me if I'm wrong.
```
                                try {
                                    Long lastCleanOffset = lastClean.get(topicPartition);
                                    OffsetsToClean offsetsToClean = cleanableOffsets(log, Optional.ofNullable(lastCleanOffset), now);
                                    // update checkpoint for logs with invalid checkpointed offsets
                                    if (offsetsToClean.forceUpdateCheckpoint) {
                                        updateCheckpoints(log.parentDirFile(), Optional.of(Map.entry(topicPartition, offsetsToClean.firstDirtyOffset)), Optional.empty());
                                    }
                                    long compactionDelayMs = maxCompactionDelay(log, offsetsToClean.firstDirtyOffset, now);
                                    preCleanStats.updateMaxCompactionDelay(compactionDelayMs);

                                    return new LogToClean(log, offsetsToClean.firstDirtyOffset,
                                            offsetsToClean.firstUncleanableDirtyOffset, compactionDelayMs > 0);
                                } catch (Throwable e) {
                                    throw new LogCleaningException(log, "Failed to calculate log cleaning stats for partition " + topicPartition, e);
                                }
```

---

@wernerdv thanks for response. you are right!